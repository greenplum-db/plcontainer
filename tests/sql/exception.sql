--  Test <defunct> processes are reaped after a new backend is created.
select pykillself();
SELECT pg_sleep(5);
-- Wait for 5 seconds so that cleanup processes exit.
\!ps -ef |grep [p]ostgres|grep defunct |wc -l
-- Then start the backend so that those <defunct> processes could be reaped.
select pyzero();
\!ps -ef |grep [p]ostgres|grep defunct |wc -l

-- Test function ok immediately after container is kill-9-ed.
select pyzero();
select pykillself();
select pyzero();

--  Test shared path write permission for unix domain socket connection.
select py_shared_path_perm();
select r_shared_path_perm();
