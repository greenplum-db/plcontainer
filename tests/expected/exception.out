-- Test trigger (not supported at this moment).
CREATE TABLE trigger_tbl (a int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE OR REPLACE FUNCTION test_trigger_func() RETURNS TRIGGER AS $$
# container: plc_python_shared
plpy.notice("trigger not supported");
$$ LANGUAGE plcontainer;
CREATE TRIGGER test_trigger AFTER INSERT ON trigger_tbl FOR EACH ROW EXECUTE PROCEDURE test_trigger_func();
INSERT INTO trigger_tbl values(0);
ERROR:  plcontainer: PL/Container does not support triggers (plcontainer.c:86)  (seg1 127.0.0.1:25433 pid=13966)
DROP TRIGGER test_trigger on trigger_tbl;
DROP TABLE trigger_tbl;
--  Test <defunct> processes are reaped after a new backend is created.
select pykillself();
ERROR:  plcontainer: Error receiving data from the client: -1. Maybe retry later. (plcontainer.c:255)
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- Wait for 5 seconds so that cleanup processes exit.
\!ps -ef |grep [p]ostgres|grep defunct |wc -l
1
-- Then start the backend so that those <defunct> processes could be reaped.
select pyzero();
 pyzero 
--------
      0
(1 row)

\!ps -ef |grep [p]ostgres|grep defunct |wc -l
0
-- Test function ok immediately after container is kill-9-ed.
select pykillself();
ERROR:  plcontainer: Error receiving data from the client: -1. Maybe retry later. (plcontainer.c:255)
select pyzero();
 pyzero 
--------
      0
(1 row)

select rkillself();
ERROR:  plcontainer: Error receiving data from the client: -1. Maybe retry later. (plcontainer.c:255)
select rint(0);
 rint 
------
    2
(1 row)

-- Test function ok immediately after container captures signal sigsegv.
select pysegvself();
ERROR:  plcontainer: Error receiving data from the client: -1. Maybe retry later. (plcontainer.c:255)
select pyzero();
 pyzero 
--------
      0
(1 row)

select rsegvself();
ERROR:  plcontainer: Error receiving data from the client: -1. Maybe retry later. (plcontainer.c:255)
select rint(0);
 rint 
------
    2
(1 row)

-- Test function ok immediately after container exits.
select pyexit();
ERROR:  plcontainer: Error receiving data from the client: -1. Maybe retry later. (plcontainer.c:255)
select pyzero();
 pyzero 
--------
      0
(1 row)

select rexit();
ERROR:  plcontainer: Error receiving data from the client: -1. Maybe retry later. (plcontainer.c:255)
select rint(0);
 rint 
------
    2
(1 row)

--  Test shared path write permission for unix domain socket connection.
select py_shared_path_perm();
ERROR:  PL/Container client exception occurred:
DETAIL:  
 Exception occurred in Python during function execution 
 Traceback (most recent call last):
  File "<string>", line 5, in py_shared_path_perm
OSError: [Errno 13] Permission denied: '/tmp/plcontainer/test_file'
select r_shared_path_perm();
 r_shared_path_perm 
--------------------
 {f}
 {t}
(2 rows)

